### PR TITLE
DCES-667: Added new endpoints to get FDC and Concor contributions for a given set of IDs.

### PR DESCRIPTION
**Pull request by vvartakMoJ**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/DCES-667)

Added 2 new endpoints viz /debt-collection-enforcement/fdc-contributions and /debt-collection-enforcement/concor-contribution-xml to get the FDC and Concor contributions data for a given list of IDs.

Found and fixed an issue whereby some fields (e.g. STATUS, JUD_APPORTION_PERCENT etc), were not mapped and hence not returned by the endpoint /debt-collection-enforcement/fdc-contribution-files.

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

## Additional checks

- Don’t forget to [run](https://github.com/ministryofjustice/laa-crimeapps-maat-functional-tests/actions/workflows/ExecuteUiTests.yaml) the MAAT functional test suite after deploying your changes to the DEV or TEST environments to ensure your changes haven’t broken any of the functional tests.

### CODE REVIEW

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java [hunk 1]
```diff
@@ -52,26 +54,24 @@ public class FdcContributionsService {
     private final DebtCollectionService debtCollectionService;
     private final FdcContributionMapper fdcContributionMapper;
 
-    private static final Function<FdcContributionsEntity, FdcContributionEntry> BUILD_FDC_ENTRY =
```
Comment (vvartakMoJ) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java:55]: This method is missing some fields like judApportionPercent, status etc which were always being returned as null. There is already a suitable mapper for this which does a better job and should be used instead.

**Review by Mhowell494 — CHANGES_REQUESTED** [review]

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java [hunk 2]
```diff
@@ -50,6 +55,21 @@ public ResponseEntity<List<ConcorContributionResponse>> concorContributionFiles(
         return ResponseEntity.ok(contributionResponses);
     }
 
+    @StandardApiResponseCodes
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java:58]: I'm not quite sure what is needed here, but happy to create a separate JIRA to look at all these things in one go as tech debt. Need to get this functionality out urgently as it's blocking other development.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/ConcorContributionsServiceTest.java [hunk 3]
```diff
@@ -352,6 +352,25 @@ void testGetConcorContributionWhenNotFound() {
         assertNull(actualResponse);
     }
 
+    @Test
+    void testGetConcorContributionXmlWhenNotFound() {
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/ConcorContributionsServiceTest.java:372]: Will add to the tech debt ticket to resolve separately.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/FdcContributionsServiceTest.java [hunk 4]
```diff
@@ -394,6 +407,31 @@ private void getFdcContributionsFile() {
...
+            .build();
+    }
+
+
+    private void assertEqualsWithExpectedValues(FdcContributionsResponse response) {
+        List<FdcContributionEntry> fdcContributionEntries = response.getFdcContributions();
+        Assertions.assertFalse(fdcContributionEntries.isEmpty());
+        FdcContributionEntry responseValue = fdcContributionEntries.get(0);
+        FdcContributionsEntity expectedValue = fdcContributionsEntityList.get(0);
+        assertEquals(expectedValue.getId(), responseValue.getId());
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/FdcContributionsServiceTest.java:427]: Will add to the tech debt ticket to resolve separately.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestControllerTest.java [hunk 5]
```diff
@@ -299,6 +303,46 @@ void testGetContributionWhenFound() throws Exception {
                 .andExpect(jsonPath("$.status").value("SENT"));
     }
 
+    @Test
+    void givenEmptyListOfIds_whenXmlIsRequested_thenBadRequestError() throws Exception {
+
+        mvc.perform(MockMvcRequestBuilders.post(String.format(ENDPOINT_URL  + CONCOR_CONTRIBUTION_XML_URL))
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .content("[]"))
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestControllerTest.java:347]: Test added

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 6]
```diff
@@ -63,11 +64,20 @@ public List<ConcorContributionResponse> getConcorContributionFiles(ConcorContrib
...
+    }
+
+    public List<ConcorContributionResponse> getConcorContributionXml(List<Integer> idList) {
+        List<ConcorContributionsEntity> concorFileList = concorRepository.findByIdIn(new HashSet(idList));
+        return getConcorContributionResponses(concorFileList);
+    }
 
+    private static List<ConcorContributionResponse> getConcorContributionResponses(
+        List<ConcorContributionsEntity> concorFileList) {
         return concorFileList.stream().map(cc -> ConcorContributionResponse.builder()
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java]: Done.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 7]
```diff
@@ -63,11 +64,20 @@ public List<ConcorContributionResponse> getConcorContributionFiles(ConcorContrib
         log.info("Searching concor contribution file with status {}, startId {} and count {}", status, concorContributionId, noOfRecords);
         Pageable pageable = PageRequest.of(0, noOfRecords, Sort.by("id"));
         final List<ConcorContributionsEntity> concorFileList = concorRepository.findByStatusAndIdGreaterThan(status, concorContributionId, pageable);
+        return getConcorContributionResponses(concorFileList);
+    }
+
+    public List<ConcorContributionResponse> getConcorContributionXml(List<Integer> idList) {
+        List<ConcorContributionsEntity> concorFileList = concorRepository.findByIdIn(new HashSet(idList));
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java]: Done

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 8]
```diff
@@ -63,11 +64,20 @@ public List<ConcorContributionResponse> getConcorContributionFiles(ConcorContrib
...
+        return getConcorContributionResponses(concorFileList);
+    }
 
+    private static List<ConcorContributionResponse> getConcorContributionResponses(
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java]: Good suggestion, done.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java [hunk 9]
```diff
@@ -52,26 +54,24 @@ public class FdcContributionsService {
...
-                    .sentenceOrderDate(Objects.nonNull(entity.getRepOrderEntity()) ? entity.getRepOrderEntity().getSentenceOrderDate() : null)
-                    .build();
-
-    public FdcContributionsResponse getFdcContributionFiles(FdcContributionsStatus status) {
-        log.info("Getting fdc contribution file with status with the -> {}", status);
-        final List<FdcContributionsEntity> fdcFileList = fdcContributionsRepository.findByStatus(status);
-        List<FdcContributionEntry> fdcContributionEntries = fdcFileList.stream()
-                .map(BUILD_FDC_ENTRY)
-                .toList();
+    private FdcContributionsResponse getFdcContributionsResponse(Supplier<List<FdcContributionsEntity>> repositoryCall) {
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java]: Good suggestion, done.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/FdcContributionsController.java [hunk 10]
```diff
@@ -115,4 +120,19 @@ public ResponseEntity<FdcContributionEntry> getFdcContribution(@PathVariable(nam
...
+    public ResponseEntity<FdcContributionsResponse> getFdcContributions(@RequestBody final List<Integer> fdcContributionIdList) {
+        log.info("Request received to get the XML for {} IDs", fdcContributionIdList.size());
+        if (fdcContributionIdList.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "ID List Empty");
+        } else if (fdcContributionIdList.size() > REQUEST_ID_LIST_SIZE_LIMIT) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Too many IDs provided, max is " + REQUEST_ID_LIST_SIZE_LIMIT);
+        } else {
+            return ResponseEntity.ok(fdcContributionsService.getFdcContributions(fdcContributionIdList));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/FdcContributionsController.java]: Good suggestion, done.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java [hunk 11]
```diff
@@ -52,26 +54,24 @@ public class FdcContributionsService {
...
+        List<FdcContributionsEntity> fdcContributionsList = repositoryCall.get();
+        List<FdcContributionEntry> fdcContributionEntries = fdcContributionsList.stream()
+            .map(fdcContributionMapper::mapFdcContribution)
+            .toList();
         return FdcContributionsResponse.builder().fdcContributions(fdcContributionEntries).build();
```
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java]: Done

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/ConcorContributionsServiceTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/ConcorContributionsServiceTest.java:372]: JIRA ticket created to handle this: https://dsdmoj.atlassian.net/browse/DCES-670

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/FdcContributionsServiceTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/service/FdcContributionsServiceTest.java:427]: JIRA ticket created: https://dsdmoj.atlassian.net/browse/DCES-670

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vvartakMoJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java]: JIRA ticket created: https://dsdmoj.atlassian.net/browse/DCES-670

**Review by Mhowell494 — APPROVED** [review]
Thank you for addressing some of the issues in this PR—it’s a good step forward. I’d like to see the remaining concerns resolved, particularly those related to the Swagger documentation, as this is quite important. If possible, please prioritise these fixes in a follow-up. That said, given the time constraints, I’m happy for this to be merged for now.

**Pull request merged**